src/main/kotlin/com/cognifide/gradle/aem/environment/ServiceAwait.kt

```diff
@@ -0,0 +1,78 @@
+package com.cognifide.gradle.aem.environment
+
+import com.cognifide.gradle.aem.common.AemExtension
+import com.cognifide.gradle.aem.common.ProgressLogger
+import com.cognifide.gradle.aem.common.Retry
+import com.cognifide.gradle.aem.common.http.HttpClient
+import com.cognifide.gradle.aem.common.http.RequestException
+import kotlin.streams.toList
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.runBlocking
+
+class ServiceAwait(private val aem: AemExtension) {
+
+    private val options = aem.environmentOptions
+    val progress = ProgressLogger.of(aem.project)
+
+    fun await() =
+            progress.launch {
+                val serviceStatuses = options.healthChecks.list.parallelStream().map { it.url to healthy(it) }.toList()
+                if (!serviceStatuses.all { it.second }) {
+                    val unavailableUrls = serviceStatuses.filter { !it.second }.map { it.first }.joinToString("\n")
+                    throw EnvironmentException("Failed to initialized all services! Following URLs are still unavailable " +
+                            "or returned different response than expected:\n$unavailableUrls")
+                }
+            }
+
+    fun await(message: String, retry: Retry, condition: () -> Boolean) {
```

Comment:
please reuse `aem.progressIndicator` / asynchronous, or use `aem.progressLogger`  / synchronous

please do not reinvent the wheel

Reply:
I need your support to reuse it. Please, let's sit together and think about how to do it. I already used some time to do it alone without satisfying effect :)